### PR TITLE
Added magic attack packet handling, pipeline, actions, hooks, and fin…

### PR DESCRIPTION
**Pull request by Twistt**
…ally plugins. Working full circle.

No damage or effects calculated yet. No sounds, animations, or projectiles yet.

### CODE REVIEW

File: src/game-engine/world/action/magic-on-npc.action.ts [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+import { ActionHook, getActionHooks } from '@engine/world/action/hooks';
+import { advancedNumberHookFilter, questHookFilter } from '@engine/world/action/hooks/hook-filters';
+import { ActionPipe, RunnableHooks } from '@engine/world/action/index';
+import { Npc } from '../actor/npc/npc';
+
+
+/**
+ * Defines a button action hook.
+ */
+export interface Magic_On_NPCActionHook extends ActionHook<Magic_On_NPCAction, magic_on_npcActionHandler> {
```
Comment (Tynarus) [inline src/game-engine/world/action/magic-on-npc.action.ts]: Only request atm is to use `MagicOnNpcActionHook` naming as opposed to using underscores in class/interface names

**Review by Tynarus — CHANGES_REQUESTED** [review]

File: src/game-engine/world/action/magic-on-npc.action.ts (hunk 1 above)
Comment (Twistt) [reply src/game-engine/world/action/magic-on-npc.action.ts]: Refactored underscores out in latest commit #13f1108

File: src/game-engine/world/actor/actor.ts [hunk 2]
```diff
@@ -36,6 +41,25 @@ export abstract class Actor {
...
     public lastMovementPosition: Position;
+    // #region Behaviors and Combat flags/checks
+    public inCombat: boolean = false;
+    public meleeDistance: number = 1;
+    public Behaviors: Behavior[] = [];
+    public isDead: boolean = false;
+    public combatTargets: Actor[] = [];
+    public hitPoints = this.skills.hitpoints.level * 4;
+    public maxHitPoints = this.skills.hitpoints.level * 4;
+    public get highestCombatSkill(): Skill {
```
Comment (Tynarus) [inline src/game-engine/world/actor/actor.ts:52]: recommend moving into `skills.ts`

File: src/game-engine/world/actor/behaviors/melee-combat.behavior.ts [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+            
+        });
+    }
+    public async doAttack():Promise<void> {
+        return new Promise<void>(resolve => {
+            //do attack stuff
+            this.Me.playAnimation(this.Me.getAttackAnimation());
+
+            const _damage = this.Me.skills.strength.level;
+            console.log(`${(this.Me as Npc).name} attacks ${(this.Them as Player).username} for ${_damage} damage! (after the CD)`);
```
Comment (Tynarus) [inline src/game-engine/world/actor/behaviors/melee-combat.behavior.ts:67]: Prefer `logger.info` but I know you were having issues with that before, so no biggie

**Review by Tynarus — APPROVED** [review]
:shipit:

**Pull request merged**